doc/developer/design/20220718_envelopes.md

```diff
@@ -0,0 +1,190 @@
+since if we crash while ingesting the initial snapshot we can just restart from zero.
+
+The main wrinkle for classic `DEBEZIUM` mode is that not only do we need to persist
+the output data, but we also need to persist the in-memory deduplication state, since it's
+not possible to recover this without re-reading the upstream topic. This will require a sidecar
+"metadata" persist shard, which we must write to while emitting the data for every timestamp.
+On restart, we will re-ingest that state as of the restart timestamp
+
+The most difficult part of the project is supporting full consistency; i.e.,
+correctly interpreting the transaction metadata topic. Currently, that topic
+cannot be deduplicated without keeping O(k) in-memory state in the number of transations
+that have ever been processed in the lifetime of the upstream database; clearly,
+this cost will be unacceptable for many upstream users. One possible approach is to
+change Debezium upstream to improve this situation; this is very feasible, but
+would mean we can't expect to support anyone but users of extremely recent Debezium versions.
+Another possibility is tolerating some amount of duplication; it's possible that we
+can still be correct while only using memory proportional to the _duplicated_ state,
+but this needs to be thought through in more detail.
+
+Another wrinkle with the transaction metadata topic is that we need custom logic for compacting it;
+briefly, we want to drop all data for transactions that have been fully processed,
+but this won't be possible if some of the transaction messages have been duplicated
+(and we only see the duplicate record after the transaction has been fully processed). Also, it has
+implications for creating new sources: if we've compacted away some past state, we won't be able to
+create new sources that are valid for those timestamps; thus we need some machinery for expressing
+that the `since` frontier of newly created sources depends on the frontier of their transaction
+topic. I spoke with Nikhil and Chae about this some and believe it's possible (though I'm
+still concerned about how it interacts with the deduplication story). At any rate it needs
+to be thought through further, which is why I've marked transaction topic support as a stretch goal.
+
+## Alternatives
+
+The main alternative is to keep the full log of upstream data in persist,
+which will obviously allow us to resume trivially (by restarting from the beginning).
+
+I advocate for rejecting this alternative because the space
+requirement and restart time requirement are unacceptable for many real-world use cases.
+
+## Open questions
+
+Robustness - bad data can cause us to end up with negatively many copies of a given row.
+We need to ensure that this is handled gracefully by all downstream operators, and
+can at worst cause Materialize to emit errors for affected sources, but not to crash.
```

Comment:
Only for the non-upsert envelope, right?

Reply:
Correct, this can't happen for `UPSERT` or `DEBEZIUM UPSERT` (unless we have bugs).

Arbitrarily weird data could cause us to OOM, though (imagine you tells us your "keys" are some giant text field). Probably that counts as "working to spec", though.